lib/ddtrace/logger.rb

```diff
@@ -10,30 +11,89 @@ class Logger < ::Logger
     def initialize(*args, &block)
       super
       self.progname = LOG_PREFIX
+
+      # Rate limit buckets
+      @buckets = {}
     end
 
     def add(severity, message = nil, progname = nil, &block)
       where = ''
+      c = nil
+      skipped_messages = ''
+
+      # Logging rate of 0 means log everything, so skip bucket checking
+      if Datadog.configuration.logging_rate > 0
+        # Get the call stack to determine who is logging
+        c = caller
```

Comment:
The goal here was to try and get as granular of a key as possible. We want to make sure that every unique log message can be logged at least once per 60 seconds.

If you have an alternative suggestion to getting this granular, definitely open to thoughts, this is the best approach I found from my limited digging (and we already were using `caller`).

Reply:
What kind of granularity do you want? What kind of information do you want as a part of the key?

Here's a copypasta from analysis I did on this sort of thing previously:

**The experiment in Ruby**

In Ruby there are 3 ways to do this:

1. `block.source_location`:
  - Works in all version of Ruby.
  - Best performance.
2. `caller.last`
  - Works in all version of Ruby.
  - Worst performance.
3. `caller_locations.last`
  - Works only in Ruby 2.0
  - Decent performance (much better than 1, a little worse than 2)

Here's what it basically looks like in practice:

```ruby
require 'benchmark'

def control
  :control
end

def control_with_block
  yield
end

def with_block(&block)
  yield
  {
    file: block.source_location.first,
    line: block.source_location.last
  }
end

def with_caller
  source_location = /(.+):(\d+):in `(.+)'/.match(caller.last)
  {
    file: source_location[1],
    line: source_location[2].to_i
  }
end

def with_caller_locations
  source_location = caller_locations.last
  {
    file: source_location.path,
    line: source_location.lineno
  }
end

puts with_block { }
puts with_caller
puts with_caller_locations
puts "\n"

n = 50000
Benchmark.bm do |x|
  x.report("control") do
    n.times do
      control
    end
  end

  x.report("control_with_block") do
    n.times do
      control_with_block do
        :control
      end
    end
  end

  x.report("block.source_location") do
    n.times do
      with_block do
        :with_block
      end
    end
  end

  x.report("caller.last") do
    n.times do
      with_caller
    end
  end

  x.report("caller_locations.last") do
    n.times do
      with_caller_locations
    end
  end
end
```

And the result:

```
{:file=>"capture.rb", :line=>35}
{:file=>"capture.rb", :line=>36}
{:file=>"capture.rb", :line=>37}

       user     system      total        real
control  0.001785   0.000000   0.001785 (  0.001781)
control_with_block  0.002680   0.000000   0.002680 (  0.002680)
block.source_location  0.027656   0.000000   0.027656 (  0.027715)
caller.last  0.396792   0.000000   0.396792 (  0.397044)
caller_locations.last  0.060642   0.000000   0.060642 (  0.060670)
```

Reply:
^ From the above you can see just how bad the use of `caller` is compared to other means.

Reply:
Ideally, I'd like to do file/line level granularity of the `log.{info,debug,warn,error,etc}` that was called.

I used `caller` because we already had that in that file, but seems like maybe refactoring to only using `caller_locations` is better, and just have `<2.0` not report `where` for error/debug and this logging to not get that granular for `<2.0`.

Seem like the best approach here?

There is an alternative approach where every log line basically also manually sets a unique key to be used: `log.warn('PAYLOAD_REJECTED', "payload #{payload} was rejected by the agent")`.

Reply:
I think using `caller_locations` and having support for 2.0+ only would be satisfactory. My only concern about using stack locations in general then is if they mutate enough with dynamic code to create an explosion in cardinality (see my other question elsewhere in this review.) I don't know if this actually a threat, just bringing it up for consideration.

The alternative approach I think from a design perspective is good because it could provide a lot of benefits (less of a performance hit, less dependent on stack composition, simplifies key cardinality, etc) but there's obviously a cost to having to manually define a key for every log message.

So I guess if you're comparing pros and cons, what approach is better? I guess it depends on what we prioritize: the original is simpler to maintain but introduces performance cost and risk at runtime, the alternative is more work to maintain but performs better and generally less risky. Honestly, I'm a bit torn and could go either way. If I had to choose, maybe the alternative because it seems simpler, has better performance, and is more customizable.

Interested in your thoughts on the above.

Maybe there's also some compromise between the two as well? As in, you can manually provide a key optionally, otherwise it tries to infer a key automatically. Then this stack signature key logic could just be a module/extension around the bucket logger. Might be a better design?